Convert DDP parameters to ReplicatedTensor during forward pass.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75753

As per the design in https://github.com/pytorch/pytorch/issues/72138,
convert DDP parameters to ReplicatedTensor during its forward pass. Concretely,
this is done as follows:

1) Create a separate `_replicated_tensor_module` which is a copy of self.module
without creating copies of the Tensors themselves.
2) Use `_replicated_tensor_module` instead of `self.module` during the forward
pass.
3) Have a context manager `_ddp_replicated_tensor` to enable this, since
certain edge cases can fail where self.module is changed out of band resulting
in discrepancy between self.module and `_replicated_tensor_module`.

Differential Revision: [D35533736](https://our.internmc.facebook.com/intern/diff/D35533736/)

pritamdamania committed:
Convert DDP parameters to ReplicatedTensor during forward pass.

As per the design in https://github.com/pytorch/pytorch/issues/72138,
convert DDP parameters to ReplicatedTensor during its forward pass. Concretely,
this is done as follows:

1) Create a separate `_replicated_tensor_module` which is a copy of self.module
without creating copies of the Tensors themselves.
2) Use `_replicated_tensor_module` instead of `self.module` during the forward
pass.
3) Have a context manager `_ddp_replicated_tensor` to enable this, since
certain edge cases can fail where self.module is changed out of band resulting
in discrepancy between self.module and `_replicated_tensor_module`.

Differential Revision: [D35533736](https://our.internmc.facebook.com/intern/diff/D35533736/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -947,6 +957,16 @@ def no_sync(self):
             self.require_backward_grad_sync = old_require_backward_grad_sync
 
+    def _run_ddp_forward(self, *inputs, **kwargs):
+        from .replicated_tensor_ddp_interop import _ddp_with_replicated_tensor_enabled
```
Do we need to do this import on every forward pass? Will it affect performance?

wanchaol replied:
yeah seems like we can just import it once in the header?

pritamdamania87 replied:
Good point, will move this up.

wanchaol commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -197,3 +198,86 @@ def test_replicated_tensor_inter_op_sharded_tensor_errors(self):
+
+        # Test with context manager.
+        from torch.nn.parallel.replicated_tensor_ddp_interop import _ddp_replicated_tensor
+        with _ddp_replicated_tensor(False):
```
why we need to test two nested context manager with False/True?

pritamdamania87 replied:
The default is `True` (set in common_distributed.py), so I'm just disabling and re-enabling to make sure the ctx manager works as expected

wanchaol commented on torch/nn/parallel/replicated_tensor_ddp_interop.py:
```diff
@@ -0,0 +1,67 @@
+        else:
+            setattr(replica, param_name, param)
+
+    for buffer_name, buffer in module._buffers.items():
```
do we also need to replicate buffers?

pritamdamania87 replied:
Keeping it simpler for now and only replicating parameter since buffer replication is actually tricky in DDP if you follow the `sync_buffers` calls, they happen only under certain conditions.

pritamdamania87 commented on torch/distributed/_shard/replicated_tensor.py:
```diff
@@ -131,3 +138,13 @@ def validate(self) -> bool:
+            from torch.distributed._shard.api import _get_current_process_group
+            self.process_group = _get_current_process_group()
+
+    def __getstate__(self):
```
No this is similar to what we do for ShardedTensor, where we use `_get_current_process_group()` above instead.

pritamdamania87 commented on torch/distributed/nn/__init__.py:
```diff
@@ -1,2 +1,4 @@
-from .api.remote_module import RemoteModule
+import torch
+if torch.distributed.rpc.is_available():
```
Yeah it is tricky how the imports are done.

pritamdamania committed:
Update on "Convert DDP parameters to ReplicatedTensor during forward pass."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
convert DDP parameters to ReplicatedTensor during its forward pass. Concretely,
this is done as follows:

1) Create a separate `_replicated_tensor_module` which is a copy of self.module
without creating copies of the Tensors themselves.
2) Use `_replicated_tensor_module` instead of `self.module` during the forward
pass.
3) Have a context manager `_ddp_replicated_tensor` to enable this, since
certain edge cases can fail where self.module is changed out of band resulting
in discrepancy between self.module and `_replicated_tensor_module`.

Differential Revision: [D35533736](https://our.internmc.facebook.com/intern/diff/D35533736/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Convert DDP parameters to ReplicatedTensor during forward pass."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
convert DDP parameters to ReplicatedTensor during its forward pass. Concretely,
this is done as follows:

1) Create a separate `_replicated_tensor_module` which is a copy of self.module
without creating copies of the Tensors themselves.
2) Use `_replicated_tensor_module` instead of `self.module` during the forward
pass.
3) Have a context manager `_ddp_replicated_tensor` to enable this, since
certain edge cases can fail where self.module is changed out of band resulting
in discrepancy between self.module and `_replicated_tensor_module`.

Differential Revision: [D35533736](https://our.internmc.facebook.com/intern/diff/D35533736/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good to me. Have only one suggestion inlined.

rohan-varma commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -197,3 +198,89 @@ def test_replicated_tensor_inter_op_sharded_tensor_errors(self):
+        self.assertEqual(model.bias, params[1])
+
+        # Validate output
+        out = ddp(replica_tensor)
```
Should we test parity with replicated tensor vs non replicated tensor?

pritamdamania87 replied:
So the existing DDP tests already do this where they end up running DDP with replicated tensor and compare with local model run.

rohan-varma commented on test/distributed/_shard/test_replicated_tensor.py:
```diff
@@ -197,3 +198,89 @@ def test_replicated_tensor_inter_op_sharded_tensor_errors(self):
+            for _ in range(5):
+                with _ddp_replicated_tensor(True):
+                    out = ddp(replica_tensor)
+                self.assertIsInstance(out, ReplicatedTensor)
```
Same comment about the parity tests, including the context manager too?

pritamdamania87 replied:
same as above

rohan-varma commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -1355,10 +1355,15 @@ def test_ddp_zero_overlap(
+        # Disable DDP + ReplicatedTensor since ZeroRedundancyOptimizer
+        # modifies the model parameters in place.
+        from torch.nn.parallel._replicated_tensor_ddp_utils import _ddp_replicated_tensor
+        with _ddp_replicated_tensor(False):
```
Curious if there are future features developed in DDP:

1. Will they be required / enforced to have replicated tensor support?
2. Will we eventually be in some sort of situation where there is a subset of features supported by replicated tensor and a subset that is not?
3. If a user is developing a feature such as one that may modify model params in place, will the developer quickly be able to realize the issue may be due to DDP running with replicated tensor and not native DDP itself?

pritamdamania87 replied:
1. Ideally we should try to ensure we can make this work and only disable support if we can't find a way around.
2. Currently there are only 2, zero redundancy optimizer and the stateless API. I think zero redundancy can actually be fixed by rebuilding the replicated_module when param.data is modified. Hopefully we can figure out a way to fix stateless too.
3. This might be tricky and not obvious to the developer why things might be failing here.

rohan-varma reviewed (approved):
LGTM, thanks! A minor question around what is expected for future developers. Also, wondering if there are any checkpointing / state_dict implications when checkpointing with replicated tensors?

zhaojuanmao reviewed (commented):
not blocking this PR, just realized that usually users call "self.module.state_dict()" to save DDP states. 

will it have issues for saving states when replicated tensor is enabled?

pritamdamania87 commented:
> will it have issues for saving states when replicated tensor is enabled?

@zhaojuanmao ReplicatedTensor is meant to be only a runtime thing (during forward pass) and is not meant to be exposed to state_dict etc. So all of that should just work as expected.

pritamdamania committed:
Update on "Convert DDP parameters to ReplicatedTensor during forward pass."


As per the design in https://github.com/pytorch/pytorch/issues/72138,
convert DDP parameters to ReplicatedTensor during its forward pass. Concretely,
this is done as follows:

1) Create a separate `_replicated_tensor_module` which is a copy of self.module
without creating copies of the Tensors themselves.
2) Use `_replicated_tensor_module` instead of `self.module` during the forward
pass.
3) Have a context manager `_ddp_replicated_tensor` to enable this, since
certain edge cases can fail where self.module is changed out of band resulting
in discrepancy between self.module and `_replicated_tensor_module`.

Differential Revision: [D35533736](https://our.internmc.facebook.com/intern/diff/D35533736/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please
